codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -232,21 +232,27 @@ private void writePendingBytes() {
         int connectionWindow = state(connectionStream).window();
 
         if (connectionWindow > 0) {
-            writeChildren(connectionStream, connectionWindow);
-            for (Http2Stream stream : connection.activeStreams()) {
-                writeChildNode(state(stream));
+            // Allocate the bytes for the connection window to the streams, but do not write.
+            allocateBytesForTree(connectionStream, connectionWindow);
+
+            // Now write all of the allocated bytes. Copying the activeStreams array to avoid
+            // side effects due to stream removal/addition which might occur as a result
+            // of end-of-stream or errors.
+            for (Http2Stream stream : connection.activeStreams().toArray(new Http2Stream[0])) {
```

Comment:
I wonder if we can avoid this copy operation. Also it seems like other scenarios where we loop over the active streams may have similar issues.  I may take a stab at https://github.com/netty/netty/pull/3537#discussion_r27249115 and see if it makes sense.